storybooks/storybook
# Theming: Theme manager with CSS custom properties

Closes #27223

## Theming 2.0

Over the years, much has changed in Storybook, but theming has remained the same. From users' perspective, the list of options felt like an in-between without a good support for multiple themes. For maintainers, it became quite a struggle to understand what colour which led to a lot of compromises and overrides. With this project, we intend to simplify the theming API for users while allowing maintainers more freedom on how to use these values.

To offer a nice transition, we introduced a function that will convert the existing theme into the new theme to make sure this work in backward compatible. Here is a quick highlight of what this update is about:
- Setting a custom theme will have to be done using pure CSS in `storybook/manager-head.html`.
- All theme values are set using CSS variables like `--sb-accent`.
- You can set a different theme for light and dark theme.
- By just updating `--sb-accent`, we automatically generate a set of colours for your text, background, icons, .. automatically.
- You can have a custom accent or background for the sidebar.
- Using class names, you can extend this API to your need and create multiple themes if you want.

RFC available here - https://github.com/storybookjs/storybook/discussions/24344
Following PR with all updated values in the manager - https://github.com/storybookjs/storybook/pull/26863

### List of CSS variables available
- `--sb-accent` - Set the accent colour everywhere.
- `--sb-background` - Set the background on the sidebar, toolbar and addon panel.
- `--sb-accentSidebar` - Set a custom accent colour for your sidebar.
- `--sb-backgroundSidebar` - Set a custom background for your sidebar.
- `--sb-documentation` - To set the colour of the docs icon on the sidebar.
- `--sb-directory` - To set the colour of the directory icon on the sidebar.
- `--sb-component` - To set the colour of the component icon on the sidebar.
- `--sb-story` - To set the colour of the story icon on the sidebar.

## Checklist for Contributors

### Testing



#### The changes in this PR are covered in the following automated tests:

#### Manual testing

1. Create a sandbox
2. Open it. Toggle between light and dak theme in the browser DevTools, see that the color changes instantly now and doesn't require a reload.
3. Add this to a `.storybook/manager-head.html` emulating a user writing a custom theme:

```html
<style>
  @media (prefers-color-scheme: light) {
    :root {
      --sb-accent: #8df68a;
    }
  }
  @media (prefers-color-scheme: dark) {
    :root {
      --sb-accent: #144d30;
    }
  }

</style>
```

4. Restart Storybook and see your new theme in effect. 💅

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_



### Documentation



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release



This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_
